#### **2. Amplitude Encoding Theory and Application Example**

Amplitude encoding is a method where classical data is represented by the amplitudes of quantum states. For example, `x` and `y` data can be encoded into a superposition state such as `a|0⟩ + b|1⟩`.

- **Technical Details:**
  - Classical data is encoded as the amplitudes of quantum states.
  - An `n`-dimensional classical data vector can be encoded into a `2^n`-dimensional quantum state.
  - The amplitudes are represented as normalized amplitudes of classical data.

- **Advantages:**
  - **Compact Encoding of Large Data Sets:** Amplitude encoding can encode large data sets compactly, increasing data density.
  - **Quantum Parallelism:** It can shorten computation times by leveraging quantum parallelism.

- **Disadvantages:**
  - **Complex Encoding Process:** Calculating and normalizing amplitudes can be complex.
  - **Low Error Tolerance:** Even a small error can significantly impact results.

- **Best Use Cases:**
  - **Large and Complex Data Sets:** Especially advantageous for processing large data sets.
  - **Machine Learning Applications:** Suitable for quantum machine learning algorithms.

- **Worst Use Cases:**
  - **Small Data Sets:** Overly complex for small data sets.
  - **Applications Sensitive to Errors:** Sensitivity to errors can be problematic in critical applications.


**Let's do an example right away.**

In [6]:
# 1.
from qiskit import QuantumCircuit,transpile
from qiskit_aer import Aer
from qiskit.visualization import plot_histogram
import matplotlib.pyplot as plt
import numpy as np

In [7]:
#2.
data_set = ['000', '011', '101', '110']

#3.
counts = [int(x, 2) for x in data_set]
norm = np.linalg.norm(counts)
normalized_data = [x / norm for x in counts]

#4.
num_qubits = int(np.log2(len(normalized_data)))
qc = QuantumCircuit(num_qubits)

#5. Amplitude Encoding
qc.initialize(normalized_data, [i for i in range(num_qubits)])

#6.
print(qc)

     ┌────────────────────────────────────────┐
q_0: ┤0                                       ├
     │  Initialize(0,0.35857,0.59761,0.71714) │
q_1: ┤1                                       ├
     └────────────────────────────────────────┘


In [8]:
#7
compiled_circuit = transpile(qc, backend=backend)
backend = Aer.get_backend('statevector_simulator')
job = backend.run(compiled_circuit, shots=1024)
result = job.result()

#8.
statevector = result.get_statevector(qc)
print("Statevector:", statevector)

Statevector: Statevector([0.        +0.j, 0.35856858+0.j, 0.5976143 +0.j,
             0.71713717+0.j],
            dims=(2, 2))


It's worth discussing the output here because there's a significant difference between what we have now and what we initially had.

The **Statevector** we obtained represents the amplitudes of all possible states in the quantum circuit. Each value represents the amplitude of a state in the quantum circuit.

To interpret these **Statevector amplitudes**:
- **0. +0.j:** The amplitude of the first state (when binary converted, it represents |00⟩) is zero. This state represents 000 in our data set.
- **0.35856858+0.j:** The amplitude of the second state (when binary converted, it represents |01⟩) is approximately 0.35856858. This state represents 011 in our data set.
- **0.5976143 +0.j:** The amplitude of the third state (when binary converted, it represents |10⟩) is approximately 0.5976143. This state represents 101 in our data set.
- **0.71713717+0.j:** The amplitude of the fourth state (when binary converted, it represents |11⟩) is approximately 0.71713717. This state represents 110 in our data set.

**So, what are these data in general?**

*These data, or normalized data, are the encoded data into the quantum circuit. These amplitudes form the superposition of quantum states.*

